src/std/foreign.ss

```diff
@@ -43,10 +44,172 @@
+      ;;
+      ;; (define-c-struct X ((a . t1) (b . t2))) =>
+      ;; In addition to the types and lambdas defined above, following additional lambdas are provided:
+      ;;
+      ;; - X-a-set!, X-b-set! setter for member variables.
+      ;;   Special compatibility for string types is provided,
+      ;;   if a string is passed as the value, then we strdup the string and set that to the
+      ;;   argument. If the struct member is already pointing to another string, then that
+      ;;   string is freed and the member will now point to a new string.
+      ;;   The cleanup of such strings are handled by the generated <struct>_ffi_free, if
+      ;;   a custom release function is provided, care should be taken while freeing.
+      ;;
+      ;;- X-a X-b accessor functions for struct members
+      (define-macro (define-c-struct struct #!optional (members '()) release-function)
+        (let* ((struct-str (symbol->string struct))
+               (struct-ptr (string->symbol (string-append struct-str "*")))
+               (shallow-ptr (string->symbol (string-append struct-str "-shallow-ptr*")))
+               (borrowed-ptr (string->symbol (string-append struct-str "-borrowed-ptr*")))
+               (string-types '(char-string nonull-char-string UTF-8-string
+                                           nonnull-UTF-8-string UTF-16-string
+                                           nonnull-UTF16-string))
+               (string-compat-required? (let loop ((m members))
+                                          (cond
+                                           ((null? m) #f)
+                                           ((member (cdr (car m)) string-types) #t)
+                                           (else (loop (cdr m))))))
+               (string-setter-body (lambda (member-name)
+                                     (let ((m (string-append "___arg1->" member-name)))
+                                       (string-append
+                                        "if(" m " == NULL)" "\n"
+                                        m "= strdup(___arg2);" "\n"
+                                        "else if (strcmp(" m ", ___arg2) != 0) {" "\n"
+                                        "free(" m ");" "\n"
+                                        m "= strdup(___arg2);" "\n"
+                                        "}" "\n"
+                                        "___return;" "\n"))))
+               (default-free-body (and string-compat-required?
+                                       (string-append
+                                        "___SCMOBJ " struct-str "_ffi_free (void *ptr) {" "\n"
+                                        "struct " struct-str " *obj = (struct " struct-str "*) ptr;" "\n"
+                                        (apply string-append
+                                          (map (lambda (m)
+                                                 (case (cdr m)
+                                                   ((char-string)
```

Comment:
This needs to catch all the string types. Maybe just `(memq  (cdr m) string-types)` instead of a case statement.